summingbird-scalding-test/src/test/scala/com/twitter/summingbird/scalding/BatchedStoreProperties.scala

```diff
@@ -0,0 +1,187 @@
+      in = arbInt.zipWithIndex.map { case (item: Int, time: Int) => (time.toLong, item) }
+      arbMap <- Arbitrary.arbitrary[Map[Int, Int]]
+      batcher = TestUtil.randomBatcher(in)
+      lastTimeStamp = in.size
+      testStore = TestStore[Int, Int]("test", batcher, arbMap, lastTimeStamp)
+    } yield (in, batcher, testStore)
+  }
+
+  implicit def arbitraryLocalMode: Arbitrary[Mode] = Arbitrary { Gen.const(Local(true)) }
+
+  property("readAfterLastBatch should return interval starting from the last batch written") = {
+    forAll {
+      (diskPipeFactory: PipeFactory[(Int, Int)],
+      interval: Intersection[InclusiveLower, ExclusiveUpper, Timestamp],
+      inputWithTimeStampAndBatcherAndStore: (List[(Long, Int)], Batcher, TestStore[Int, Int]),
+      mode: Mode) =>
+        val (inputWithTimeStamp, batcher, testStore) = inputWithTimeStampAndBatcherAndStore
+        val result = testStore.readAfterLastBatch(diskPipeFactory)((interval, mode))
+
+        result match {
+          case Right(((Intersection(InclusiveLower(readIntervalLower), ExclusiveUpper(_)), _), _)) => {
+            //readInterval should start from the last written interval in the store
+            val start: Timestamp = batcher.earliestTimeOf(testStore.initBatch.next)
+            implicitly[Ordering[Timestamp]].equiv(readIntervalLower, start)
+          }
+          case Right(_) => false
+          case Left(_) => interval == Empty()
+        }
+    }
+  }
+
+  property("readAfterLastBatch should not extend the end of interval requested") = {
+    forAll {
+      (diskPipeFactory: PipeFactory[(Int, Int)],
+      interval: Intersection[InclusiveLower, ExclusiveUpper, Timestamp],
+      inputWithTimeStampAndBatcherAndStore: (List[(Long, Int)], Batcher, TestStore[Int, Int]),
+      mode: Mode) =>
+        val (inputWithTimeStamp, batcher, testStore) = inputWithTimeStampAndBatcherAndStore
+        val result = testStore.readAfterLastBatch(diskPipeFactory)((interval, mode))
+
+        result match {
+          case Right(((Intersection(InclusiveLower(_), ExclusiveUpper(readIntervalUpper)), _), _)) => {
+            //readInterval should start from the last written interval in the store
+            implicitly[Ordering[Timestamp]].lteq(readIntervalUpper, interval.upper.upper)
+          }
+          case Right(_) => false
+          case Left(_) => interval == Empty()
+        }
+    }
+  }
+
+  property("the end of merged interval is never extended") = {
+    forAll {
+      (diskPipeFactory: PipeFactory[(Int, Int)],
+      interval: Intersection[InclusiveLower, ExclusiveUpper, Timestamp],
+      inputWithTimeStampAndBatcherAndStore: (List[(Long, Int)], Batcher, TestStore[Int, Int]),
+      mode: Mode) =>
+        val (inputWithTimeStamp, batcher, testStore) = inputWithTimeStampAndBatcherAndStore
+        val mergeResult = testStore.merge(diskPipeFactory, implicitly[Semigroup[Int]], com.twitter.summingbird.option.Commutative, 10)((interval, mode))
```

Comment:
let's make an Arbitrary Commutativity and test for both Commutative and NonCommutative